### DISCUSSION TITLE
[Q&A] Can I implement asynchronous algorithm on NVFlare? HOW?

### DISCUSSION THREAD

**Discussion by Eric716** [NVIDIA/NVFlare#2405] [Q&A] [answered]
### Python version (`python3 -V`)

3.8

### NVFlare version (`python3 -m pip list | grep "nvflare"`)

2.3.0

### Operating system

Ubuntu 20.04

### Have you successfully run any of the following examples?

- [ ] hello-numpy-sag with simulator
- [X] hello-pt with simulator
- [ ] hello-numpy-sag with POC
- [X] hello-pt with POC

### Please describe your question

Can I implement asynchronous algorithm on NVFlare?
I want to implement asynchronous algorithm like: FedAsync... on NVFlare. 
Does anyone have any suggestions or thoughts? Thank you very much.

**Comment by YuanTingHsieh** [comment]
@holgerroth can you help comment on this? thanks

**Comment by holgerroth** [comment, marked as answer]
Thanks for the question @Eric716. Yes, NVFlare can supports asynchronous FL. For example, our controllers provide both blocking and non-blocking versions to send training Tasks to clients, see `broadcast` and `broadcast_and_wait` [here](https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fnvflare.readthedocs.io%2Fen%2Fmain%2Fapidocs%2Fnvflare.apis.impl.controller.html&data=05%7C01%7Chroth%40nvidia.com%7Ce96ff56414224fcebe4a08dbf21117d4%7C43083d15727340c1b7db39efd9ccc17a%7C0%7C0%7C638369927475134896%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=Pd3ySLd8Q%2FEi3GPA7QiP%2F47Uv377F90q8qEaaFqZc10%3D&reserved=0). The “wait” version is used in synchronous FL workflows, e.g., in [ScatterAndGather](https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fnvflare.readthedocs.io%2Fen%2Fmain%2Fapidocs%2Fnvflare.app_common.workflows.scatter_and_gather.html%23nvflare.app_common.workflows.scatter_and_gather.ScatterAndGather&data=05%7C01%7Chroth%40nvidia.com%7Ce96ff56414224fcebe4a08dbf21117d4%7C43083d15727340c1b7db39efd9ccc17a%7C0%7C0%7C638369927475134896%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=XWBYcrD20dY6y8hagH5ymKH6A6%2B7syUPXeJYLKeSRe8%3D&reserved=0). Without wait, you could use it to implement asynchronous FL.
 
Every ‘[Task’](https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fnvflare.readthedocs.io%2Fen%2Fmain%2Fapidocs%2Fnvflare.apis.controller_spec.html%23nvflare.apis.controller_spec.Task&data=05%7C01%7Chroth%40nvidia.com%7Ce96ff56414224fcebe4a08dbf21117d4%7C43083d15727340c1b7db39efd9ccc17a%7C0%7C0%7C638369927475134896%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=TQYBytTL7S2mCX3lZk7yoyJmI%2BNephRrNtH66VJog9o%3D&reserved=0) that’s been sent to a client has a `result_received_cb` that can be used that to do some operation on the server side, i.e. add the weight updates to the current global model.

**Reply by Eric716** [reply]
Thanks! Is there any example for that?

**Reply by holgerroth** [reply]
We don't have an asynchronous training example at this time. However, you can take a look at this [cross-site evaluation workflow](https://github.com/NVIDIA/NVFlare/blob/385dc3c2e4c797340a484e6fc7af719cc7622dff/nvflare/app_common/workflows/cross_site_model_eval.py#L405) to see the asynchronous communication pattern in action. If you end up implementing an example, please consider contributing it back to the repo.
